compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeLowering.cpp

```diff
@@ -354,6 +354,413 @@ struct DmaWaitToTctSyncConverter final
   }
 };
 
+//===----------------------------------------------------------------------===//
+// AIEDeviceBuilder utilities
+//===----------------------------------------------------------------------===//
+// TODO(avarma): Copied from LowerToAIE. Templatize it later because of a few
+//               nuances.
+using BDDimLayoutAndLength = std::pair<AMDAIE::BDDimLayoutArrayAttr, int64_t>;
+
+BDDimLayoutAndLength convertSizeStrideToBDDimLayoutArrayAttr(
+    IRRewriter &rewriter, ArrayRef<int64_t> sizes, ArrayRef<int64_t> strides) {
+  assert(sizes.size() == strides.size() &&
+         "expected stride and size vectors of same size");
+  SmallVector<AMDAIE::BDDimLayoutAttr, 4> bdDimLayoutAttr;
+  // If the access pattern (strides/sizes) have a single dimension, make it
+  // implicit with an empty `BDDimLayoutAttr` as this is what the AIE dialect
+  // expects.
+  if (strides.size() == 1 && strides[0] == 1) {
+    return std::make_pair(AMDAIE::BDDimLayoutArrayAttr::get(
+                              rewriter.getContext(), ArrayRef(bdDimLayoutAttr)),
+                          sizes[0]);
+  }
+  bdDimLayoutAttr.reserve(sizes.size());
+  // Compute the length of the DMA transfer.
+  int64_t transferLength =
+      sizes.empty()
+          ? 0
+          : std::accumulate(sizes.begin(), sizes.end(), 1, std::multiplies<>());
+  for (auto [size, stride] : llvm::zip(sizes, strides)) {
+    bdDimLayoutAttr.push_back(
+        AMDAIE::BDDimLayoutAttr::get(rewriter.getContext(), size, stride));
+  }
+  return std::make_pair(AMDAIE::BDDimLayoutArrayAttr::get(
+                            rewriter.getContext(), ArrayRef(bdDimLayoutAttr)),
+                        transferLength);
+}
+
+/// Create a new `amdaie.dma_start` op with a sequence of DMA BD blocks.
+///
+/// Example of a S2MM DMA start op being created with two DMA blocks performing
+/// a circular double buffering DMA operation:
+///
+///  %0 = amdaie.dma_start(%tile_0_1, S2MM, 0) {
+///       amdaie.use_lock(%lock_0_1_51, AcquireGreaterOrEqual(2))
+///       amdaie.dma_bd(%buffer_0_1_49 : memref<2048xi32, 1 : i32>) {len = 2048
+///       : i32} amdaie.use_lock(%lock_0_1_52, Release(2)) amdaie.next_bd ^bb1
+///     ^bb1:  // pred: ^bb2
+///       amdaie.use_lock(%lock_0_1_51, AcquireGreaterOrEqual(2))
+///       amdaie.dma_bd(%buffer_0_1_50 : memref<2048xi32, 1 : i32>) {len = 2048
+///       : i32} amdaie.use_lock(%lock_0_1_52, Release(2)) amdaie.next_bd ^bb2
+///     ^bb2:
+///       amdaie.end
+///  }
+LogicalResult createDMABlocks(
+    IRRewriter &rewriter, Operation *tileOp, AMDAIE::DMAChannelDir channelDir,
+    int channelIndex, ArrayRef<int64_t> sizes, ArrayRef<int64_t> strides,
+    size_t acqNum, size_t relNum, int64_t offset,
+    const SmallVector<AMDAIE::BufferOp> &bufferOps,
+    const std::pair<AMDAIE::LockOp, AMDAIE::LockOp> &locks,
+    std::optional<uint8_t> pktId) {
+  OpBuilder::InsertionGuard g(rewriter);
+
+  // Create DMA channel.
```

Comment:
I know this is copied, but maybe update this comment to "Create DMA start on channel".